apps/dav/lib/CalDAV/Reminder/NotificationProvider/EmailProvider.php

```diff
@@ -0,0 +1,501 @@
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+namespace OCA\DAV\CalDAV\Reminder\NotificationProvider;
+
+use DateTime;
+use DateTimeImmutable;
+use OCP\IConfig;
+use OCP\IL10N;
+use OCP\ILogger;
+use OCP\IURLGenerator;
+use OCP\L10N\IFactory as L10NFactory;
+use OCP\Mail\IEMailTemplate;
+use OCP\Mail\IMailer;
+use OCP\IUser;
+use Sabre\VObject\Component\VEvent;
+use Sabre\VObject;
+use Sabre\VObject\Parameter;
+use Sabre\VObject\Property;
+
+/**
+ * Class EmailProvider
+ *
+ * @package OCA\DAV\CalDAV\Reminder\NotificationProvider
+ */
+class EmailProvider extends AbstractProvider {
+
+	/** @var string */
+	public const NOTIFICATION_TYPE = 'EMAIL';
+
+	/** @var IMailer */
+	private $mailer;
+
+	/**
+	 * @param IConfig $config
+	 * @param IMailer $mailer
+	 * @param ILogger $logger
+	 * @param L10NFactory $l10nFactory
+	 * @param IUrlGenerator $urlGenerator
+	 */
+	public function __construct(IConfig $config,
+								IMailer $mailer,
+								ILogger $logger,
+								L10NFactory $l10nFactory,
+								IURLGenerator $urlGenerator) {
+		parent::__construct($logger, $l10nFactory, $urlGenerator, $config);
+		$this->mailer = $mailer;
+	}
+
+	/**
+	 * Send out notification via email
+	 *
+	 * @param VEvent $vevent
+	 * @param string $calendarDisplayName
+	 * @param array $users
+	 * @throws \Exception
+	 */
+	public function send(VEvent $vevent,
+						 string $calendarDisplayName,
+						 array $users=[]):void {
+		$fallbackLanguage = $this->getFallbackLanguage();
+
+		$emailAddressesOfSharees = $this->getEMailAddressesOfAllUsersWithWriteAccessToCalendar($users);
+		$emailAddressesOfAttendees = $this->getAllEMailAddressesFromEvent($vevent);
+
+		// Quote from php.net:
+		// If the input arrays have the same string keys, then the later value for that key will overwrite the previous one.
+		// => if there are duplicate email addresses, it will always take the system value
+		$emailAddresses = array_merge(
+			$emailAddressesOfAttendees,
+			$emailAddressesOfSharees
+		);
+
+		$sortedByLanguage = $this->sortEMailAddressesByLanguage($emailAddresses, $fallbackLanguage);
+		$organizer = $this->getOrganizerEMailAndNameFromEvent($vevent);
+
+		foreach($sortedByLanguage as $lang => $emailAddresses) {
+			if ($this->hasL10NForLang($lang)) {
```

Comment:
Shouldn't this check be negated?

```suggestion
			if (!$this->hasL10NForLang($lang)) {
```

Reply:
Indeed, let me fix that.